experimental/segmenter/src/math_helper.rs

```diff
@@ -2,59 +2,338 @@
+        self.data.iter_mut().for_each(|v| {
+            *v = v.exp() / sm;
+        });
+    }
+}
+
+impl<'a> MatrixBorrowed<'a, 1> {
+    #[allow(dead_code)] // could be useful
+    pub fn dot_1d(&self, other: MatrixBorrowed<1>) -> f32 {
+        debug_assert_eq!(self.dims, other.dims);
+        unrolled_dot(self.data, other.data)
+    }
+}
+
+impl<'a> MatrixBorrowedMut<'a, 1> {
+    /// Calculate the dot product of a and b, adding the result to self.
+    ///
+    /// Note: For better dot product efficiency, if `b` is MxN, then `a` should be N;
+    /// this is the opposite of standard practice.
+    pub fn add_dot_2d(&mut self, a: MatrixBorrowed<1>, b: MatrixBorrowed<2>) {
+        let m = a.dim();
+        let n = self.as_borrowed().dim();
+        debug_assert_eq!(m, b.dim().1);
+        debug_assert_eq!(n, b.dim().0);
+        for i in 0..n {
+            if let (Some(dest), Some(b_sub)) = (self.as_mut_slice().get_mut(i), b.submatrix::<1>(i))
+            {
+                *dest += unrolled_dot(a.data, b_sub.data);
+            } else {
+                debug_assert!(false);
+            }
+        }
+    }
+}
+
+impl<'a> MatrixBorrowedMut<'a, 2> {
+    /// Calculate the dot product of a and b, adding the result to self.
+    ///
+    /// Self should be _MxN_; `a`, _O_; and `b`, _MxNxO_.
+    pub fn add_dot_3d(&mut self, a: MatrixBorrowed<1>, b: MatrixBorrowed<3>) {
+        let m = a.dim();
+        let n = self.as_borrowed().dim().0 * self.as_borrowed().dim().1;
+        debug_assert_eq!(m, b.dim().2);
+        debug_assert_eq!(n, b.dim().0 * b.dim().1);
+        // Note: The following two loops are equivalent, but the second has more opportunity for
+        // vectorization since it allows the vectorization to span submatrices.
+        // for i in 0..b.dim().0 {
+        //     self.submatrix_mut::<1>(i).add_dot_2d(a, b.submatrix(i));
+        // }
+        let lhs = a.as_slice();
+        for i in 0..n {
+            if let (Some(dest), Some(rhs)) = (
+                self.as_mut_slice().get_mut(i),
+                b.as_slice().get(i * m..(i + 1) * m),
+            ) {
+                *dest += unrolled_dot(lhs, rhs);
+            } else {
+                debug_assert!(false);
+            }
+        }
+    }
+}
+
+// Polyfill float operations with libm in case we're no_std.
+#[allow(unused_imports)]
+use num_traits::Float;
+
+/// `tanh` computes the tanh function for a scalar value.
+#[inline]
+pub fn tanh(x: f32) -> f32 {
+    x.tanh()
+}
+
+/// `sigmoid` computes the sigmoid function for a scalar value.
+#[inline]
+pub fn sigmoid(x: f32) -> f32 {
+    1.0 / (1.0 + (-x).exp())
+}
+
+/// Compute the dot product.
+///
+/// `xs` and `ys` must be the same length
+///
+/// (From ndarray 0.15.6)
+#[allow(clippy::indexing_slicing)] // all indexing is protected by the entry assertion
+fn unrolled_dot(xs: &[f32], ys: &[f32]) -> f32 {
+    debug_assert_eq!(xs.len(), ys.len());
+    if xs.len() != ys.len() {
+        return 0.0;
```

Comment:
> I'm not happy with fallibility in this PR. There are many places where you could define exact preconditions that are probably always satisfied, so you wouldn't need fallibility, however there's a lot `Option` propagation, usually around indexing. I think that should be avoided because (a) it should be possible to reason that indexing is correct and (b) recovering from a `None` with `0.0`, `String::new()`, or `Self::empty()` seems logically wrong.

We could labor over avoiding option propagation with lots of comments in code that need to be reviewed and verified, but it's much easier to prove that this is non-panicky code by not making assumptions and just propagating default values, as suggested in #2768

Reply:
FYI: There's an earlier version of this (one of the intermediate commits) that used lots of panicky indexing operations. After I changed them to option propagation, there was no measurable difference in performance.

Reply:
> it's much easier to prove that this is non-panicky code by not making assumptions and just propagating default values 

For the goal of not panicking, sure it's way easier to use `Option`s and just pull zeroes out of thin air if you encounter a `None`, but I don't think that's correct. If this happens it's essentially logic-flaw-in-garbage out, but we do want to detect these logic flaws if they're there. 

Also, zero is not a good default value for doing (matrix) multiplications. There isn't really one, but iiuc these cases are intended to never be hit, so we shouldn't need to define values for them. 

> We could labor over avoiding option propagation with lots of comments in code that need to be reviewed and verified

I believe we should. We do this everywhere else where we use indexing, I don't think we use logic like `.get(x).unwrap_or_default()` anywhere to not go through the work of showing that `x` is valid.

Reply:
I believe there is now exactly 1 place where we inject GIGO values, which is the top level `word_segmenter` function. I strongly prefer to keep this code style of propagating `Option::None`. It's easier to read, easier to write, easier to reason about, more robust to future changes, consistent with the new style guide, and just as efficient as the alternative.

Reply:
This is like putting a `try { ... } catch (Throwable t) { ... }` at the top level, which I consider an anti-pattern. It's unclear what the error cases are, and it makes the code harder to understand and to maintain.

Looking through the code, most `None`s come from indexing. If you verify the bounds in the constructor you can consider indexing infallible in the actual logic, and eliminate the garbage-out.